File: java/app/src/main/java/com/jossy/android/mobilebenchmarkappjava/activity/CpuTest.java
package com.jossy.android.mobilebenchmarkappjava.activity;

import android.annotation.SuppressLint;
import android.util.Log;

import com.jossy.android.mobilebenchmarkappjava.CPUTest;
import com.jossy.android.mobilebenchmarkappjava.R;
import com.jossy.android.mobilebenchmarkappjava.consts.Config;
import com.jossy.android.mobilebenchmarkappjava.data.CpuResult;
import com.jossy.android.mobilebenchmarkappjava.data.TestResult;

public class CpuTest extends BaseTestActivity {

    private static final String TAG = "CpuTest";

    @Override
    protected void initializeActivity() {
        setContentView(R.layout.activity_cpu_test);
        statusText = findViewById(R.id.cpuStatus);
    }

    @Override
    protected void onDestroy() {
        CPUTest.shutdown();
        super.onDestroy();
    }

    @Override
    protected void executeBenchmark() throws Exception {
        CPUTest.initialize(null);
        initializeCsvWriter();

        for (int i = 0; i < Config.sampleCount; i++) {
            executeSingleIteration(i);
        }

        if (csvWriter != null)
            csvWriter.flush();
        CPUTest.shutdown();
    }

    private void executeSingleIteration(int index) {
        long start = System.currentTimeMillis();
        CpuResult result = CPUTest.runBenchmarkIterations(Config.cpuIterations, null);
        long duration = System.currentTimeMillis() - start;

        TestResult testResult = new TestResult(
                "CPU Test",
                duration,
                "threads=" + result.threads,
                true);
        logTestResult(index, testResult);
        updateProgress(index + 1);
    }

    @Override
    @SuppressLint("SetTextI18n")
    protected String getProgressDisplayText(int currentIteration) {
        return "CPU Test: " + currentIteration + " / " + Config.sampleCount;
    }

    @Override
    protected String getTestName() {
        return "CPU Test";
    }

    @Override
    protected void initializeCsvWriter() throws Exception {
        if (csvWriter != null)
            return;

        java.io.File file = new java.io.File(csvPath != null ? csvPath : getDefaultCsvPath());
        csvWriter = new com.jossy.android.mobilebenchmarkappjava.io.BufferedCsvWriter(file, Config.bufferSize,
                64 * 1024, "CPU Test",
                new com.jossy.android.mobilebenchmarkappjava.io.BufferedCsvWriter.CsvFormatter() {
                    @Override
                    public String format(com.jossy.android.mobilebenchmarkappjava.data.TestEntry entry) {
                        return entry.iteration + "," + entry.result.getExecutionTimeMs() + "\n";
                    }

                    @Override
                    public String getHeader() {
                        return "iteration,elapsedTimeMs\n";
                    }
                });
        csvWriter.initialize();
    }
}

File: kotlin/app/src/main/kotlin/com/jossy/android/mobilebenchmarkappjava/activity/CpuTest.kt
package com.jossy.android.mobilebenchmarkappjava.activity

import android.widget.TextView
import android.widget.LinearLayout
import android.graphics.Color
import com.jossy.android.mobilebenchmarkappjava.CPUTest
import com.jossy.android.mobilebenchmarkappjava.consts.Config
import com.jossy.android.mobilebenchmarkappjava.data.TestResult

class CpuTest : BaseTestActivity() {

    override fun initializeActivity() {
        // Create layout programmatically (no R resource needed)
        val layout = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
            )
            setPadding(32, 32, 32, 32)
            setBackgroundColor(Color.WHITE)
        }
        statusText = TextView(this).apply {
            text = "CPU Test: Starting..."
            textSize = 18f
            setTextColor(Color.BLACK)
            layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )
        }
        layout.addView(statusText)
        setContentView(layout)
    }

    override fun executeBenchmark() {
        initializeCsvWriter()
        repeat(Config.sampleCount) { i ->
            executeSingleIteration(i)
        }
        csvWriter?.flush()
    }

    private fun executeSingleIteration(index: Int) {
        val start = System.currentTimeMillis()
        val result = CPUTest.runBenchmarkIterations(Config.cpuIterations.toLong(), null)
        val duration = System.currentTimeMillis() - start

        val testResult = TestResult(
            testName = "CPU Test",
            executionTimeMs = duration,
            details = "threads=${result.threads}",
            success = true
        )
        logTestResult(index, testResult)
        updateProgress(index + 1)
    }

    override fun getProgressDisplayText(currentIteration: Int): String =
        "CPU Test: $currentIteration / ${Config.sampleCount}"

    override fun getTestName(): String = "CPU Test"
}
